File: Adidas/adidas-fliteTrakr/java/src/main/java/com/nestis/adidas/fliteTrakr/model/strategy/StopsStrategy.java
package com.nestis.adidas.fliteTrakr.model.strategy;

import java.util.List;

/**
 * Defines a strategy to get the number of routes that comply with the number of stops permitted.
 * #7: How many different connections with maximum 3 stops exists between NUE and FRA?
 * #8: How mand different connections with exactly 1 stop exists between LHR and AMS?
 * @author nestis
 *
 */
public class StopsStrategy extends FlexibleRouteTemplate {
	
	/**
	 * Stops number.
	 */
	private Integer stops; 
	
	/**
	 * Boolean logic of the stops number. Maximum, minimum or exactly.
	 */
	private StopsStrategyEnum type;
	
	public StopsStrategy(String origin, String dest, Integer stops, StopsStrategyEnum type) {
		super(origin, dest);
		this.stops = stops;
		this.type = type;
	}

	@Override
	public String getResult(List<Route> routeFound) {
		if (routeFound.isEmpty()) {
			return "No such connection found!";
		} else {
			long trips = routeFound.stream()
				.filter(r -> {
 					boolean valid = false;
 					// We always substract 2, because origin and destination doens't count as stop overs.
					if (StopsStrategyEnum.EXACTLY == type) {
						valid = r.route.size() - 2 == stops;
					} else if (StopsStrategyEnum.MAXIMUM == type) {
						valid = r.route.size() - 2 <= stops;
					} else {
						valid = r.route.size() - 2 >= stops;
					}
					return valid;
				}).count();
			
			return String.valueOf(trips);
		}
	}

}

File: Adidas/adidas-fliteTrakr/js/src/model/strategy/stopsStrategy.js
const RouteTemplate = require('./routeTemplate');
const Constants = require('../../util/constants');

/**
 * Defines the behaviour that gets the number of routes that comply with the number of stops permitted.
 * How many different connections with maximum 3 stops exists between NUE and FRA?
 * How mand different connections with exactly 1 stop exists between LHR and AMS?
 * @param {*} origin Origin airport.
 * @param {*} dest Destination airport.
 * @param {*} stops Number of stops.
 * @param {*} type Boolean logic for the stop. Must be a value for Constants.VALUES.*
 */
function StopsStrategy(origin, dest, stops, type) {
    RouteTemplate.call(this, origin, dest);
    this.stops = stops;
    this.type = type;
}
StopsStrategy.prototype = Object.create(RouteTemplate.prototype);
StopsStrategy.prototype.constructor = StopsStrategy;

/**
 * Returns the info from the routes obtained.
 */
StopsStrategy.prototype.getInfo = function() {
	if (this.routeFound.length > 0) {
        let routes = [];
        switch (this.type) {
            case Constants.VALUES.EXACTLY:
                routes = this.routeFound.filter(r => r.path.length - 2 === this.stops);
                break;
            case Constants.VALUES.MINIMUM:
                routes = this.routeFound.filter(r => r.path.length - 2 >= this.stops);
                break;
            case Constants.VALUES.MAXIMUM:
                routes = this.routeFound.filter(r => r.path.length - 2 <= this.stops);
                break;
        }
        return routes.length;
	} else {
		return 'No such route found!';
	}
}

module.exports = StopsStrategy;
